tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter.cc

```diff
@@ -277,5 +293,223 @@ StatusOr<bool> CudnnFusedConvRewriter::Run(HloModule* module) {
+      return ConvWithConvertOrClamp{instr, gte, conv};
+    }
+  }
+  return absl::nullopt;
+}
+
+StatusOr<bool> TryRewriteForConvertOrClampImpl(ConvWithConvertOrClamp match) {
+  auto conv = match.conv;
+  auto gte = match.gte;
+  auto convert_or_clamp = match.convert_or_clamp;
+
+  // Change type on conv and gte
+  auto convert_out_type = convert_or_clamp->shape().element_type();
+  conv->mutable_shape()->mutable_tuple_shapes(0)->set_element_type(
+      convert_out_type);
+  gte->mutable_shape()->set_element_type(convert_out_type);
+
+  // Remove clamp/convert and so on and just keep
+  // get_tuple_element(custom_call(x,w, ...))
+  convert_or_clamp->ReplaceAllUsesWithDifferentShape(gte);
+  conv->parent()->RemoveInstructionAndUnusedOperands(convert_or_clamp);
+
+  return true;
+}
+
+StatusOr<bool> TryRewriteForFinalOutput(ConvWithConvertOrClamp match) {
+  bool changed = false;
+  // When the matched clamp has a single user, which is convert<int8>, we
+  // will absorb it, if
+  // 1. the side_input matches a convert<float>(int8_side_input), or
+  // 2. there is no side input
+  const auto is_one_to_one_X_to_Y_cast = [](const HloInstruction* instr,
+                                            PrimitiveType X,
+                                            PrimitiveType Y) -> bool {
+    return (instr->opcode() == HloOpcode::kConvert &&
+            instr->shape().element_type() == Y && instr->operand_count() == 1 &&
+            instr->operand(0)->user_count() == 1 &&
+            instr->operand(0)->shape().element_type() == X);
+  };
+
+  if (match.conv->operand_count() < 4) {
+    // Conv input #3 (zero based) is side_input, after x, w, and bias.
+    // Side input doesn't exist in this case.
+    TF_ASSIGN_OR_RETURN(changed, TryRewriteForConvertOrClampImpl(match));
+  } else if (is_one_to_one_X_to_Y_cast(match.conv->operand(3), S8, F32)) {
+    // If side_input has a convert_float_to_int8, absorb it as well.
+    auto side_converter = match.conv->mutable_operand(3);
+    side_converter->ReplaceAllUsesWithDifferentShape(
+        side_converter->mutable_operand(0));
+    side_converter->parent()->RemoveInstructionAndUnusedOperands(
+        side_converter);
+
+    TF_ASSIGN_OR_RETURN(changed, TryRewriteForConvertOrClampImpl(match));
+  }
+  return changed;
+}
+
+// Fuse the clamp/convert pattern with the int8 convolution custom call
+// (either pure or fused) for int8 output
+StatusOr<bool> RunFuseClamp(HloModule* module) {
+  bool changed = false;
+  for (HloComputation* computation : module->MakeNonfusionComputations()) {
+    std::vector<ConvWithConvertOrClamp> matches;
+    int num_forward_convs = 0;
+    for (auto instr : computation->instructions()) {
+      auto match = FindConvWithClampAndConvertToInt8(instr);
+      if (match.has_value()) {
+        matches.push_back(*match);
+      }
+    }
+    std::vector<std::pair<HloInstruction*, std::unique_ptr<HloInstruction>>>
+        replacements;
```

Comment:
The replacements vector is not actually used?

Reply:
No, it is not.  I should have cleaned it up.

Reply:
replacements in this function has been removed.